public/app/listeners/executeAlgorithm.js

```diff
@@ -0,0 +1,140 @@
+const fs = require('fs');
+const path = require('path');
+const logger = require('../logger');
+const { RESULTS_FOLDER } = require('../config/config');
+const { createNewDataset } = require('./loadDataset');
+const {
+  DATASETS_COLLECTION,
+  RESULTS_COLLECTION,
+  ALGORITHMS_COLLECTION,
+  EXECUTIONS_COLLECTION,
+} = require('../db');
+const { EXECUTE_ALGORITHM_CHANNEL } = require('../../shared/channels');
+const {
+  PROGRAMMING_LANGUAGES,
+  EXECUTION_STATUSES,
+} = require('../config/config');
+const executePythonAlgorithm = require('./executePythonAlgorithm');
+const {
+  UNKNOWN_PROGRAMMING_LANGUAGE,
+  ERROR_EXECUTION_PROCESS,
+  ERROR_GENERAL,
+} = require('../../shared/errors');
+const {
+  EXECUTE_ALGORITHM_SUCCESS,
+  EXECUTE_ALGORITHM_ERROR,
+} = require('../../shared/types');
+
+const createNewResultDataset = ({
+  name,
+  filepath,
+  algorithmId,
+  description,
+}) => {
+  const result = createNewDataset({
+    name,
+    filepath,
+    folderPath: RESULTS_FOLDER,
+    description,
+  });
+  result.algorithmId = algorithmId;
+  return result;
+};
+
+const executeAlgorithm = (mainWindow, db) => (
+  event,
+  { datasetId, algorithmId, id: executionId },
+) => {
+  const channel = `${EXECUTE_ALGORITHM_CHANNEL}_${executionId}`;
+  try {
+    // get corresponding dataset
+    const { filepath, name: datasetName, description } = db
+      .get(DATASETS_COLLECTION)
+      .find({ id: datasetId })
+      .value();
+
+    // get the corresponding algorithm
+    const {
+      filepath: algorithmFilepath,
+      name: algorithmName,
+      language,
+    } = db.get(ALGORITHMS_COLLECTION).find({ id: algorithmId }).value();
+
+    const tmpPath = path.join(RESULTS_FOLDER, `tmp_${executionId}.json`);
+
+    // prepare success callback function
+    // copy tmp as new result dataset
+    // update execution and return it
+    const onSuccess = () => {
```

Comment:
ok I've been trying to debug an issue but maybe it's more efficient to collaborate on it :)
There is something off I suspect in both or either this `onSuccess` callback and/or `process.on('close', ...)` in `executePythonAlgorithm`. specifically... try executing an algorithm as normal (just once). if you look at `db.json`, even if after the algorithm has executed successfully, its status remains `"pending"`. HOWEVER, if you run another algorithm, the **previous** algorithm becomes marked as `"status": "success"`. but as a result this is quite buggy because if you exit the app and reopen it then an algorithm previously showing as complete now shows as pending.
so something is off, my guess is in this function or in `executePythonAlgorithm`

Reply:
I am not sure about the bug you are experiencing. Is it related to the db from the developer tab? I got some inaccuracies with it when working on Graasp Desktop. We can go through it together someday, but my guess would go for an deprecated value at some time.

Reply:
no, doesn't have to do with the db from the developer tab. it can be reproduced on ceaeccf exactly as I described in my comment. in any case, the changes you've made in `executeAlgorithm.js` in the last commit fix this error: specifically, you now set an execution as running (lines 51-54), and in the `onSuccess` callback you `.write()` the update to `EXECUTIONS_COLLECTION` instead of reading it (with `.value()`). maybe you caught/fixed the error independently of my comment... in any case, it is fixed